storage/remote/queue_manager.go

```diff
@@ -846,6 +917,9 @@ func (s *shards) sendSamples(ctx context.Context, samples []prompb.TimeSeries, b
 	if err != nil {
 		level.Error(s.qm.logger).Log("msg", "non-recoverable error", "count", len(samples), "err", err)
 		s.qm.metrics.failedSamplesTotal.Add(float64(len(samples)))
+		if time.Since(begin) > s.qm.flushDeadline {
```

Comment:
This means that any retries that take longer than the flushDeadline will cause a hard shutdown? Can we just use the context cancel to stop sending?

Reply:
> This means that any retries that take longer than the flushDeadline will cause a hard shutdown?

Yes

> Can we just use the context cancel to stop sending?

Sorry, do you mean to just skip the current send or to stop the shard? The current implementation here stops the reshard by using the context's cancel (hardShutdown is the cacncelFunc). If not, then I think I did not get your point. Can you please explain a bit more?

Reply:
I will rephrase, right now `(*shards).stop()` contains all of the logic around soft and hard shutdowns. It would be ideal if that logic would still only be contained in `stop()`. It is not obvious to be calling `hardShutdown` from inside a method called `sendSamples`, and I am sure that will lead to unexpected behavior in the future.

I think that ends up meaning that 
```
	if resharding {
		return
	}
```
inside of `stop()` should be removed. Since the shard would be sending to the new queue instead of to the remote, shutdown should still happen pretty quickly. You may have to setup some background processing in reshardLoop to avoid blocking on stop, but that is preferable IMO.

Reply:
This has been addressed. I think this can be resolved now.